Firestore/core/test/unit/core/target_test.cc

```diff
@@ -0,0 +1,385 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/core/target.h"
+
+#include <cmath>
+
+#include "Firestore/core/src/core/bound.h"
+#include "Firestore/core/src/core/query.h"
+#include "Firestore/core/src/model/document_set.h"
+#include "Firestore/core/src/model/field_path.h"
+#include "Firestore/core/src/model/mutable_document.h"
+#include "Firestore/core/src/model/value_util.h"
+#include "Firestore/core/src/nanopb/message.h"
+#include "Firestore/core/test/unit/testutil/testutil.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace core {
+
+namespace {
+
+using core::Bound;
+using firebase::firestore::util::ComparisonResult;
+using model::CanonicalId;
+using model::DocumentComparator;
+using model::Equals;
+using model::FieldIndex;
+using model::FieldPath;
+using model::MutableDocument;
+using model::Segment;
+using nanopb::MakeSharedMessage;
+
+using testing::AssertionResult;
+using testing::Not;
+using testutil::Array;
+using testutil::BlobValue;
+using testutil::CollectionGroupQuery;
+using testutil::DbId;
+using testutil::Doc;
+using testutil::Field;
+using testutil::Filter;
+using testutil::MakeFieldIndex;
+using testutil::Map;
+using testutil::OrderBy;
+using testutil::Query;
+using testutil::Ref;
+using testutil::Value;
+using testutil::Vector;
+
+void VerifyBound(Bound bound,
+                 bool inclusive,
+                 std::vector<google_firestore_v1_Value> values) {
+  EXPECT_EQ(inclusive, bound.inclusive());
+  EXPECT_EQ(values.size(), bound.position()->values_count);
+  for (size_t i = 0; i < values.size(); ++i) {
+    const auto& expected_value = values[i];
+    EXPECT_TRUE(Equals(expected_value, bound.position()->values[i]))
+        << "Values should be equal: Expected: " << CanonicalId(expected_value)
+        << ", Actual: " << CanonicalId(bound.position()->values[i]) << "";
+  }
+}
+
+TEST(TargetTest, EmptyQueryBound) {
+  Target target = testutil::Query("c").ToTarget();
+  FieldIndex index = MakeFieldIndex("c");
+
+  auto lower_bound = target.GetLowerBound(index);
+  ASSERT_TRUE(lower_bound.has_value());
+  VerifyBound(lower_bound.value(), true, {});
```

Comment:
Can VerifyBound take an optional, so that we can move the `HasValue` check?

Reply:
Done.